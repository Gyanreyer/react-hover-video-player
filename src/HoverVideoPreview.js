import { ClassNames } from '@emotion/core';
import React from 'react';

/**
 * @typedef   {object}  VideoSource
 * @property  {string}  src - The src URL string to use for a video player source
 * @property  {string}  type - The media type of the video, ie 'video/mp4'
 */

/**
 * @typedef   {object}  VideoCaptionsTrack
 * @property  {string}  src - The src URL string for the captions track file
 * @property  {string}  srcLang - The language code for the language that these captions are in
 * @property  {string}  label - The title of the captions track
 */

/**
 * @component HoverVideoPreview
 *
 * @param {node}   previewOverlay - The contents to render over the video while it's not playing
 * @param {!(string|string[]|VideoSource|VideoSource[])}  videoSrc -
 *                                                Source(s) to use for the video player. Accepts 3 different formats:
 *                                                - **String**: the URL string to use as the video player's src
 *                                                - **Object**: an object with an `src` attribute defining the src URL and a `type` attribute for the source's media type (ie, 'video/mp4')
 *                                                - **Array**: if you would like to provide multiple sources, you can provide an array of URL strings and/or objects with the shape described above
 * @param {!(string|string[]|VideoCaptionsTrack|VideoCaptionsTrack[])}
 * @param {bool}    [shouldRestartOnVideoStopped=true] - Whether the video should reset to the beginning every time it stops playing after the user mouses out of the preview
 * @param {number}  [fadeTransitionDuration=400] - The transition duration in ms for how long it should take for the overlay to fade in/out
 * @param {func}    [onStartingVideo] - Optional callback for every time the user mouses over or focuses on the hover preview and we attempt to start the video
 * @param {func}    [onStartedVideo] - Optional callback for when the video has been successfully started
 * @param {func}    [onStoppingVideo] - Optional callback for every time the user mouses out or blurs the hover preview and we attempt to stop the video
 * @param {func}    [onStoppedVideo] - Optional callback for when the video has successfully been stopped
 * @param {bool}    [shouldShowVideoControls=false] - Whether the video player should show the browser's controls
 * @param {bool}    [shouldVideoLoop=true] - Whether the video player should loop when it reaches the end
 * @param {bool}    [isVideoMuted=true] - Whether the video player should be muted
 * @param {string}  [videoPreload='metadata'] - What the video should preload. Possible values:
 *                                              - 'none': No part of the video will be preloaded before we try to play it
 *                                              - 'metadata' (default): Only the video's metadata will be preloaded, including information such as its length
 *                                              - 'auto': The entire video will be preloaded even if it may not be played
 * @param {string}  [className] - Optional className to apply custom styling to the container element
 * @param {string}  [overlayWrapperClassName] - Optional className to apply custom styling to the overlay contents' wrapper
 * @param {string}  [videoClassName] - Optional className to apply custom styling to the video element
 * @param {object}  [style] - Style object to apply custom CSS styles to the hover preview container
 */
const HoverVideoPreview = React.forwardRef(
  (
    {
      previewOverlay,
      videoSrc,
      videoCaptions,
      shouldRestartOnVideoStopped = true,
      overlayFadeTransitionDuration = 400,
      onStartingVideo,
      onStartedVideo,
      onStoppingVideo,
      onStoppedVideo,
      shouldShowVideoControls = false,
      shouldVideoLoop = true,
      isVideoMuted = true,
      videoPreload = 'metadata',
      className = '',
      overlayWrapperClassName = '',
      videoClassName = '',
      style = null,
    },
    ref
  ) => {
    const [isOverlayHidden, setIsOverlayHidden] = React.useState(false);

    const videoRef = React.useRef();
    const pauseVideoTimeoutRef = React.useRef();
    const playPromiseRef = React.useRef();

    // Parse the `videoSrc` prop into an array of VideoSource objects to be used for the video player
    const parsedVideoSources = React.useMemo(() => {
      if (videoSrc == null) {
        // A videoSrc value is required in order to make the video player work
        console.error(
          "Error: 'videoSrc' prop is required for HoverVideoPreview component"
        );

        return [];
      }

      return (
        // Make sure we can treat the videoSrc value as an array
        []
          .concat(videoSrc)
          // Parse our video source values into an array of VideoSource objects that can be used to render sources for the video
          .reduce((sourceArray, source) => {
            if (typeof source === 'string') {
              // If the source is a string, it's an src URL so format it into a VideoSource object and add it to the array
              sourceArray.push({ src: source });
            } else if (source && source.src) {
              // If the source is an object with an src, just add it to the array
              sourceArray.push({ src: source.src, type: source.type });
            } else {
              // Log an error if one of the videoSrc values is invalid
              console.error(
                "Error: invalid value provided to HoverVideoPreview prop 'videoSrc':",
                source
              );
            }

            return sourceArray;
          }, [])
      );
    }, [videoSrc]);

    const parsedVideoCaptions = React.useMemo(() => {
      // If no captions were provided, return an empty array
      if (!videoCaptions) return [];

      return (
        // Make sure we can treat the videoSrc value as an array
        []
          .concat(videoCaptions)
          // Parse our video captions values into an array of VideoCaptionsTrack
          // objects that can be used to render caption tracks for the video
          .reduce((captionsArray, captions) => {
            if (typeof captions === 'string') {
              captionsArray.push({ src: captions });
            } else if (captions && captions.src) {
              captionsArray.push({
                src: captions.src,
                srcLang: captions.srcLang,
                label: captions.label,
              });
            } else {
              // Log an error if one of the videoCaptions values is invalid
              console.error(
                "Error: invalid value provided to HoverVideoPreview prop 'videoCaptions'"
              );
            }

            return captionsArray;
          }, [])
      );
    }, [videoCaptions]);

    /**
     * @function  attemptStopVideo
     *
     * Stops the video and fades the preview overlay in when the user mouses out from or blurs the video container element
     */
    const attemptStopVideo = () => {
      if (onStoppingVideo) onStoppingVideo();

      const stopVideo = () => {
        // Start fading the overlay back in to cover up the video before it's paused
        setIsOverlayHidden(false);

        // If we already had a timeout going to pause the video, cancel it so we can
        // replace it with a new one
        clearTimeout(pauseVideoTimeoutRef.current);

        // Set a timeout with the duration of the overlay's fade transition so the video
        // won't stop until it's fully hidden
        pauseVideoTimeoutRef.current = setTimeout(
          () => {
            videoRef.current.pause();

            if (shouldRestartOnVideoStopped) {
              // If we should restart the video, reset its time to the beginning
              videoRef.current.currentTime = 0;
            }

            if (onStoppedVideo) onStoppedVideo();
          },
          previewOverlay ? overlayFadeTransitionDuration : 0
        );
      };

      if (playPromiseRef.current instanceof Promise) {
        // If we have a promise from when we attempted to start playing the video,
        // ensure we wait to pause until the play promise is completed so we don't interrupt it
        playPromiseRef.current.finally(stopVideo);
      } else {
        // If we don't have a play attempt promise, just go ahead and pause
        stopVideo();
      }
    };

    /**
     * @function  attemptStartVideo
     *
     * Starts the video and fades the preview overlay out when the user mouses over or focuses in the video container element
     */
    const attemptStartVideo = () => {
      if (onStartingVideo) onStartingVideo();

      // If the user quickly moved their mouse away and then back over the container,
      // cancel any outstanding timeout that would pause the video
      clearTimeout(pauseVideoTimeoutRef.current);

      if (videoRef.current.paused) {
        // If the video is not playing already, start playing it
        // Keep a reference to the returned promise for the play attempt so we can avoid interrupting
        // it when we need to pause the video
        playPromiseRef.current = videoRef.current.play().then(() => {
          setIsOverlayHidden(true);
          if (onStartedVideo) onStartedVideo();
        });
      } else {
        // If the video is already playing, just make sure we keep the overlay hidden
        setIsOverlayHidden(true);
      }
    };

    return (
      <ClassNames>
        {({ css, cx }) => (
          <div
            onMouseEnter={attemptStartVideo}
            onFocus={attemptStartVideo}
            onMouseOut={attemptStopVideo}
            onBlur={attemptStopVideo}
            className={cx(
              css`
                position: relative;
              `,
              className
            )}
            style={style}
            ref={ref}
          >
            {previewOverlay && (
              <div
                className={cx(
                  css`
                    /* Overlay wrapper should completely cover the dimensions of the video */
                    position: absolute;
                    top: 0;
                    bottom: 0;
                    left: 0;
                    right: 0;
                    width: 100%;
                    height: 100%;

                    /* Fade overlay in/out based on whether it should be visible or hidden */
                    opacity: ${isOverlayHidden ? 0 : 1};
                    transition: opacity ${overlayFadeTransitionDuration}ms;
                  `,
                  overlayWrapperClassName
                )}
              >
                {previewOverlay}
              </div>
            )}
            {/* eslint-disable-next-line jsx-a11y/media-has-caption */}
            <video
              controls={shouldShowVideoControls}
              loop={shouldVideoLoop}
              muted={isVideoMuted}
              playsInline
              preload={videoPreload}
              ref={videoRef}
              className={cx(
                css`
                  width: 100%;
                  display: block;
                `,
                videoClassName
              )}
            >
              {parsedVideoSources.map(({ src, type }) => (
                <source key={src} src={src} type={type} />
              ))}
              {parsedVideoCaptions.map(({ src, srcLang, label }) => (
                <track
                  kind="captions"
                  src={src}
                  srcLang={srcLang}
                  label={label}
                />
              ))}
            </video>
          </div>
        )}
      </ClassNames>
    );
  }
);

export default HoverVideoPreview;
