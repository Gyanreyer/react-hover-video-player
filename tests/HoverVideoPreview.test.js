import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import HoverVideoPreview from '../src';

const expectVideoHasCorrectAttributes = (
  videoElement,
  { muted = true, loop = true, controls = false, preload = 'metadata' } = {}
) => {
  expect(videoElement).not.toBeNull();

  if (muted) {
    expect(videoElement).toHaveAttribute('muted');
  } else {
    expect(videoElement).not.toHaveAttribute('muted');
  }

  if (loop) {
    expect(videoElement).toHaveAttribute('loop');
  } else {
    expect(videoElement).not.toHaveAttribute('loop');
  }

  if (controls) {
    expect(videoElement).toHaveAttribute('controls');
  } else {
    expect(videoElement).not.toHaveAttribute('controls');
  }

  expect(videoElement).toHaveAttribute('preload', preload);

  expect(videoElement).toHaveAttribute('playsInline');
};

describe('HoverVideoPreview', () => {
  describe('Handles valid videoSrc prop values correctly', () => {
    test('correctly handles receiving a string for the videoSrc prop', () => {
      const { container } = render(
        <HoverVideoPreview videoSrc="/fake/video-file.mp4" />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(1);
      expect(videoSources[0]).toHaveAttribute('src', '/fake/video-file.mp4');
      expect(videoSources[0]).not.toHaveAttribute('type');
    });

    test('correctly handles receiving an array of strings for the videoSrc prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc={['/fake/video-file.webm', '/fake/video-file.mp4']}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(2);
      expect(videoSources[0]).toHaveAttribute('src', '/fake/video-file.webm');
      expect(videoSources[0]).not.toHaveAttribute('type');
      expect(videoSources[1]).toHaveAttribute('src', '/fake/video-file.mp4');
      expect(videoSources[1]).not.toHaveAttribute('type');
    });

    test('correctly handles receiving a valid object for the videoSrc prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc={{ src: '/fake/video-file.mp4', type: 'video/mp4' }}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(1);
      expect(videoSources[0]).toHaveAttribute('src', '/fake/video-file.mp4');
      expect(videoSources[0]).toHaveAttribute('type', 'video/mp4');
    });

    test('correctly handles receiving an array of objects for the videoSrc prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc={[
            { src: '/fake/video-file.webm', type: 'video/webm' },
            { src: '/fake/video-file.mp4', type: 'video/mp4' },
          ]}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(2);
      expect(videoSources[0]).toHaveAttribute('src', '/fake/video-file.webm');
      expect(videoSources[0]).toHaveAttribute('type', 'video/webm');
      expect(videoSources[1]).toHaveAttribute('src', '/fake/video-file.mp4');
      expect(videoSources[1]).toHaveAttribute('type', 'video/mp4');
    });

    test('correctly handles receiving an array with a mix of strings and objects for the videoSrc prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc={[
            '/fake/video-file.webm',
            { src: '/fake/video-file.avi' },
            { src: '/fake/video-file.mp4', type: 'video/mp4' },
          ]}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(3);
      expect(videoSources[0]).toHaveAttribute('src', '/fake/video-file.webm');
      expect(videoSources[0]).not.toHaveAttribute('type');
      expect(videoSources[1]).toHaveAttribute('src', '/fake/video-file.avi');
      expect(videoSources[1]).not.toHaveAttribute('type');
      expect(videoSources[2]).toHaveAttribute('src', '/fake/video-file.mp4');
      expect(videoSources[2]).toHaveAttribute('type', 'video/mp4');
    });
  });

  describe('Handles invalid videoSrc prop values correctly', () => {
    let originalConsoleError;

    beforeEach(() => {
      // Mock the console.error function so we can verify that an error was logged correctly
      originalConsoleError = console.error;
      console.error = jest.fn();
    });

    afterEach(() => {
      console.error = originalConsoleError;
    });

    test('correctly handles not receiving a videoSrc prop', () => {
      const { container } = render(<HoverVideoPreview />);

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(0);

      // Should have logged an error warning that the 'videoSrc' prop is required
      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error).toHaveBeenCalledWith(
        "Error: 'videoSrc' prop is required for HoverVideoPreview component"
      );
    });

    test('correctly handles receiving a single invalid value for the videoSrc prop', () => {
      const { container } = render(<HoverVideoPreview videoSrc={100} />);

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(0);

      // Should have logged an error warning that the 'videoSrc' prop is required
      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error).toHaveBeenCalledWith(
        "Error: invalid value provided to HoverVideoPreview prop 'videoSrc':",
        100
      );
    });

    test('correctly handles receiving an invalid value in an array for the videoSrc prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc={[
            'valid-video-file.webm',
            false,
            { src: 'valid-video-file.mp4', type: 'video/mp4' },
          ]}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoSources = videoElement.querySelectorAll('source');
      expect(videoSources).toHaveLength(2);
      expect(videoSources[0]).toHaveAttribute('src', 'valid-video-file.webm');
      expect(videoSources[0]).not.toHaveAttribute('type');
      expect(videoSources[1]).toHaveAttribute('src', 'valid-video-file.mp4');
      expect(videoSources[1]).toHaveAttribute('type', 'video/mp4');

      // Should have logged an error warning that the 'videoSrc' prop is required
      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error).toHaveBeenCalledWith(
        "Error: invalid value provided to HoverVideoPreview prop 'videoSrc':",
        false
      );
    });
  });

  describe('Handles valid videoCaptions prop values correctly', () => {
    test('correctly handles receiving a string for the videoCaptions prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc="/fake/video-file.mp4"
          videoCaptions="/fake/captions-file.vtt"
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoTracks = videoElement.querySelectorAll('track');
      expect(videoTracks).toHaveLength(1);
      expect(videoTracks[0]).toHaveAttribute('kind', 'captions');
      expect(videoTracks[0]).toHaveAttribute('src', '/fake/captions-file.vtt');
    });

    test('correctly handles receiving a valid object for the videoCaptions prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc="/fake/video-file.mp4"
          videoCaptions={{
            src: '/fake/captions-file-en.vtt',
            srcLang: 'en',
            label: 'English',
          }}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoTracks = videoElement.querySelectorAll('track');
      expect(videoTracks).toHaveLength(1);
      expect(videoTracks[0]).toHaveAttribute('kind', 'captions');
      expect(videoTracks[0]).toHaveAttribute(
        'src',
        '/fake/captions-file-en.vtt'
      );
      expect(videoTracks[0]).toHaveAttribute('srclang', 'en');
      expect(videoTracks[0]).toHaveAttribute('label', 'English');
    });

    test('correctly handles receiving an array of objects for the videoCaptions prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc="/fake/video-file.mp4"
          videoCaptions={[
            {
              src: '/fake/captions-file-en.vtt',
              srcLang: 'en',
              label: 'English',
            },
            {
              src: '/fake/captions-file-fr.vtt',
              srcLang: 'fr',
              label: 'French',
            },
            {
              src: '/fake/captions-file-de.vtt',
              srcLang: 'de',
              label: 'German',
            },
          ]}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoTracks = videoElement.querySelectorAll('track');
      expect(videoTracks).toHaveLength(3);
      expect(videoTracks[0]).toHaveAttribute('kind', 'captions');
      expect(videoTracks[0]).toHaveAttribute(
        'src',
        '/fake/captions-file-en.vtt'
      );
      expect(videoTracks[0]).toHaveAttribute('srclang', 'en');
      expect(videoTracks[0]).toHaveAttribute('label', 'English');

      expect(videoTracks[1]).toHaveAttribute('kind', 'captions');
      expect(videoTracks[1]).toHaveAttribute(
        'src',
        '/fake/captions-file-fr.vtt'
      );
      expect(videoTracks[1]).toHaveAttribute('srclang', 'fr');
      expect(videoTracks[1]).toHaveAttribute('label', 'French');

      expect(videoTracks[2]).toHaveAttribute('kind', 'captions');
      expect(videoTracks[2]).toHaveAttribute(
        'src',
        '/fake/captions-file-de.vtt'
      );
      expect(videoTracks[2]).toHaveAttribute('srclang', 'de');
      expect(videoTracks[2]).toHaveAttribute('label', 'German');
    });
  });

  describe('Handles invalid videoCaptions prop values correctly', () => {
    let originalConsoleError;

    beforeEach(() => {
      // Mock the console.error function so we can verify that an error was logged correctly
      originalConsoleError = console.error;
      console.error = jest.fn();
    });

    afterEach(() => {
      console.error = originalConsoleError;
    });

    test('correctly handles receiving a single invalid value for the videoCaptions prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc="fake/video-file.mp4"
          videoCaptions={false}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoTracks = videoElement.querySelectorAll('track');
      expect(videoTracks).toHaveLength(0);

      // Should have logged an error warning that the 'videoSrc' prop is required
      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error).toHaveBeenCalledWith(
        "Error: invalid value provided to HoverVideoPreview prop 'videoCaptions'",
        false
      );
    });

    test('correctly handles receiving an invalid value in an array for the videoCaptions prop', () => {
      const { container } = render(
        <HoverVideoPreview
          videoSrc="fake/video-file.mp4"
          videoCaptions={[
            {
              src: '/fake/captions-file-en.vtt',
              srcLang: 'en',
              label: 'English',
            },
            false,
            100,
            {
              src: '/fake/captions-file-fr.vtt',
              srcLang: 'fr',
              label: 'French',
            },
          ]}
        />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement);

      const videoTracks = videoElement.querySelectorAll('track');
      expect(videoTracks).toHaveLength(2);
      expect(videoTracks[0]).toHaveAttribute('kind', 'captions');
      expect(videoTracks[0]).toHaveAttribute(
        'src',
        '/fake/captions-file-en.vtt'
      );
      expect(videoTracks[0]).toHaveAttribute('srclang', 'en');
      expect(videoTracks[0]).toHaveAttribute('label', 'English');

      expect(videoTracks[1]).toHaveAttribute('kind', 'captions');
      expect(videoTracks[1]).toHaveAttribute(
        'src',
        '/fake/captions-file-fr.vtt'
      );
      expect(videoTracks[1]).toHaveAttribute('srclang', 'fr');
      expect(videoTracks[1]).toHaveAttribute('label', 'French');

      // Should have logged an error warning that the 'videoSrc' prop is required
      expect(console.error).toHaveBeenCalledTimes(2);
      expect(console.error).toHaveBeenNthCalledWith(
        1,
        "Error: invalid value provided to HoverVideoPreview prop 'videoCaptions'",
        false
      );
      expect(console.error).toHaveBeenNthCalledWith(
        2,
        "Error: invalid value provided to HoverVideoPreview prop 'videoCaptions'",
        100
      );
    });
  });

  describe('Handles video props correctly', () => {
    test('isVideoMuted prop correctly sets muted attribute on video', () => {
      const { container, rerender } = render(
        <HoverVideoPreview videoSrc="fake/video-file.mp4" />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement, { muted: true });

      // Re-render with the video unmuted
      rerender(
        <HoverVideoPreview
          videoSrc="fake/video-file.mp4"
          isVideoMuted={false}
        />
      );
      expectVideoHasCorrectAttributes(videoElement, { muted: false });
    });

    test('shouldShowVideoControls prop correctly sets controls attribute on video', () => {
      const { container, rerender } = render(
        <HoverVideoPreview videoSrc="fake/video-file.mp4" />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement, { controls: false });

      // Re-render with controls enabled on the video
      rerender(
        <HoverVideoPreview
          videoSrc="fake/video-file.mp4"
          shouldShowVideoControls
        />
      );
      expectVideoHasCorrectAttributes(videoElement, { controls: true });
    });

    test('shouldVideoLoop prop correctly sets loop attribute on video', () => {
      const { container, rerender } = render(
        <HoverVideoPreview videoSrc="fake/video-file.mp4" />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement, { loop: true });

      // Re-render with looping disabled on the video
      rerender(
        <HoverVideoPreview
          videoSrc="fake/video-file.mp4"
          shouldVideoLoop={false}
        />
      );
      expectVideoHasCorrectAttributes(videoElement, { loop: false });
    });

    test('videoPreload prop correctly sets preload attribute on video', () => {
      const { container, rerender } = render(
        <HoverVideoPreview videoSrc="fake/video-file.mp4" />
      );

      expect(container).toMatchSnapshot();

      const videoElement = container.querySelector('video');
      expectVideoHasCorrectAttributes(videoElement, { preload: 'metadata' });

      // Re-render with an updated video preload value
      rerender(
        <HoverVideoPreview videoSrc="fake/video-file.mp4" videoPreload="auto" />
      );
      expectVideoHasCorrectAttributes(videoElement, { preload: 'auto' });
    });
  });
});
