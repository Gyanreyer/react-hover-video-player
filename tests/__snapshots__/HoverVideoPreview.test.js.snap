// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverVideoPreview Handles invalid videoCaptions prop values correctly correctly handles receiving a single invalid value for the videoCaptions prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles invalid videoCaptions prop values correctly correctly handles receiving an invalid value in an array for the videoCaptions prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
      <track
        kind="captions"
        label="English"
        src="/fake/captions-file-en.vtt"
        srclang="en"
      />
      <track
        kind="captions"
        label="French"
        src="/fake/captions-file-fr.vtt"
        srclang="fr"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles invalid videoSrc prop values correctly correctly handles not receiving a videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    />
  </div>
</div>
`;

exports[`HoverVideoPreview Handles invalid videoSrc prop values correctly correctly handles receiving a single invalid value for the videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    />
  </div>
</div>
`;

exports[`HoverVideoPreview Handles invalid videoSrc prop values correctly correctly handles receiving an invalid value in an array for the videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="valid-video-file.webm"
      />
      <source
        src="valid-video-file.mp4"
        type="video/mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoCaptions prop values correctly correctly handles receiving a string for the videoCaptions prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.mp4"
      />
      <track
        kind="captions"
        src="/fake/captions-file.vtt"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoCaptions prop values correctly correctly handles receiving a valid object for the videoCaptions prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.mp4"
      />
      <track
        kind="captions"
        label="English"
        src="/fake/captions-file-en.vtt"
        srclang="en"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoCaptions prop values correctly correctly handles receiving an array of objects for the videoCaptions prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.mp4"
      />
      <track
        kind="captions"
        label="English"
        src="/fake/captions-file-en.vtt"
        srclang="en"
      />
      <track
        kind="captions"
        label="French"
        src="/fake/captions-file-fr.vtt"
        srclang="fr"
      />
      <track
        kind="captions"
        label="German"
        src="/fake/captions-file-de.vtt"
        srclang="de"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoSrc prop values correctly correctly handles receiving a string for the videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoSrc prop values correctly correctly handles receiving a valid object for the videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.mp4"
        type="video/mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoSrc prop values correctly correctly handles receiving an array of objects for the videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.webm"
        type="video/webm"
      />
      <source
        src="/fake/video-file.mp4"
        type="video/mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoSrc prop values correctly correctly handles receiving an array of strings for the videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.webm"
      />
      <source
        src="/fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles valid videoSrc prop values correctly correctly handles receiving an array with a mix of strings and objects for the videoSrc prop 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="/fake/video-file.webm"
      />
      <source
        src="/fake/video-file.avi"
      />
      <source
        src="/fake/video-file.mp4"
        type="video/mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles video props correctly isVideoMuted prop correctly sets muted attribute on video 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles video props correctly shouldShowVideoControls prop correctly sets controls attribute on video 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles video props correctly shouldVideoLoop prop correctly sets loop attribute on video 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`HoverVideoPreview Handles video props correctly videoPreload prop correctly sets preload attribute on video 1`] = `
<div>
  <div
    class="css-8hvobu"
  >
    <video
      class="css-q5gg9"
      loop=""
      muted=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;
